--- a/pppd/ain.c	2017-05-17 18:43:35.225672421 +0800
+++ b/pppd/main.c	2017-05-25 16:21:15.092470385 +0800
@@ -318,6 +318,7 @@
     struct protent *protp;
     char numbuf[16];
 
+    strlcpy(path_chapfile, _PATH_CHAPFILE, sizeof(path_chapfile));
     strlcpy(path_ipup, _PATH_IPUP, sizeof(path_ipup));
     strlcpy(path_ipdown, _PATH_IPDOWN, sizeof(path_ipdown));
     strlcpy(path_ipv6up, _PATH_IPV6UP, sizeof(path_ipv6up));

--- a/pppd/pppd.h	2017-05-17 18:43:35.221672467 +0800
+++ b/pppd/pppd.h	2017-05-25 16:24:42.045797270 +0800
@@ -325,6 +325,7 @@
 extern int	connect_delay;	/* Time to delay after connect script */
 extern int	max_data_rate;	/* max bytes/sec through charshunt */
 extern int	req_unit;	/* interface unit number to use */
+extern char     path_chapfile[MAXPATHLEN]; /* pathname of chap-secrets */
 extern char	path_ipup[MAXPATHLEN]; /* pathname of ip-up script */
 extern char	path_ipdown[MAXPATHLEN]; /* pathname of ip-down script */
 extern char	path_ipv6up[MAXPATHLEN]; /* pathname of ipv6-up script */


--- a/pppd/options.c	2017-05-17 18:43:35.257672055 +0800
+++ b/pppd/options.c	2017-05-25 16:30:35.257342002 +0800
@@ -116,6 +116,7 @@
 bool	tune_kernel;		/* may alter kernel settings */
 int	connect_delay = 1000;	/* wait this many ms after connect script */
 int	req_unit = -1;		/* requested interface unit */
+char    path_chapfile[MAXPATHLEN]; /* pathname of chap-secrets */
 char	path_ipup[MAXPATHLEN];	/* pathname of ip-up script */
 char	path_ipdown[MAXPATHLEN];/* pathname of ip-down script */
 char	path_ipv6up[MAXPATHLEN];	/* pathname of ipv6-up script */
@@ -315,6 +316,10 @@
       "Unset user environment variable",
       OPT_A2PRINTER | OPT_NOPRINT, (void *)user_unsetprint },
 
+    { "chap-secrets", o_string, path_chapfile,
+      "Set pathname of chap-secrets",
+      OPT_PRIV|OPT_STATIC, NULL, MAXPATHLEN },
+
     { "ip-up-script", o_string, path_ipup,
       "Set pathname of ip-up script",
       OPT_PRIV|OPT_STATIC, NULL, MAXPATHLEN },

--- a/pppd/auth.c	2017-05-25 16:33:51.046995855 +0800
+++ b/pppd/auth.c	2017-05-25 16:33:27.823270946 +0800
@@ -1641,7 +1641,7 @@
 	}
     }
 
-    filename = _PATH_CHAPFILE;
+    filename = path_chapfile;
     f = fopen(filename, "r");
     if (f == NULL)
 	return 0;
@@ -1736,7 +1736,7 @@
 	    return 0;
 	}
     } else {
-	filename = _PATH_CHAPFILE;
+	filename = path_chapfile;
 	addrs = NULL;
 	secbuf[0] = 0;

